---
title: "Homework 6 (Week 7)"
author: "Zeyu Wang/Charlie"
date: "February 24, 2019"
output:
  html_document: default
  pdf_document: default
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

```

# Report

### Question 1 

#####All of my featureCount files were generated by commands below:

```{Codes}
Codes:

# From class:
mkdir counting
cd counting
ln -s /home/frd2007/ANGSD_2019/alignment
spack load subread@1.6.2
featureCounts -a /home/frd2007/ANGSD_2019/RNA-seq/refGenome_S_cerevisiae/sacCer3.gtf -o featureCounts_Gierlinski.txt alignment/*bam

# Further commands for exon:
featureCounts -f -t exon -O -a /home/frd2007/ANGSD_2019/RNA-seq/refGenome_S_cerevisiae/sacCer3.gtf -o featureCounts_exon_Gierlinski.txt alignment/*bam

# Move to R
scp zew4001@pascal.med.cornell.edu:ANGSD/counting/*.summary .

```

#### Result
##### After reading manual of featureCount and searching on internet, what I got was still not too complex. I understand that commands like "--minOverlap", "--fracOverlap" can change the minimum value of a successful overlapping required, also "-Q" can change the minimum value of quality score of a read. But I am more interested in looking for differences between featureCount results of gene and exon in same condition, that was the reason why I used a simpler command this time.
##### Here I only add three parameters to generate the result of exon: 
* First, command "-f" changes the whole perform of featureCount from gene to exon. 
* Second, command "-t" will only find exons in annotation, which make sense since I was doing exon featureCount. 
* At last, command "-O" seems to be a really important part, especially for STAR result, it says that STAR would not dealing with reads which overlap with multiple genes or exons, so when we featureCount them, we need to add this option manually, which can make output better.


### Question 2
#### Plot for gene featureCount
```{r, fig.align='center'}

library(ggplot2)
library(tidyr)
library(dplyr)

gene <- read.delim("C:\\Users\\WZY77\\Study\\WCM\\Courses\\ANGSD\\ANGSD\\featureCounts_Gierlinski.txt.summary", row.names=1) %>%
  t %>%
  as.data.frame %>%
  mutate(sample=gsub("_Aligned.sortedByCoord.out.bam", "", rownames(.)))

gene <- gather(gene,Assign, Reads, -sample)
gene <- gene[c(which(gene$Reads!=0)),]

ggplot(data=gene, aes(x=sample, y=Reads, fill=Assign)) +
  geom_bar(stat="identity", color="black", position=position_dodge())+
  theme_minimal() + coord_flip()

```

#### Plot for exon featureCount
```{r, fig.align='center'}

exon <- read.delim("C:\\Users\\WZY77\\Study\\WCM\\Courses\\ANGSD\\ANGSD\\featureCounts_exon_Gierlinski.txt.summary", row.names=1) %>%
  t %>%
  as.data.frame %>%
  mutate(sample=gsub("_Aligned.sortedByCoord.out.bam", "", rownames(.)))

exon <- gather(exon,Assign, Reads, -sample)
exon <- exon[c(which(exon$Reads!=0)),]

ggplot(data=exon, aes(x=sample, y=Reads, fill=Assign)) +
  geom_bar(stat="identity", color="black", position=position_dodge()) + theme_minimal() + coord_flip()

```

###Question 3

#####Actually the most different part of these graph should be number of bars for each example, which is the result of using command "-O" in exon featureCount, a option deals with the reads that overlaps with multiple exons. However, I also feel it is kind of strange that after using "-O", all of reads that were unassigned due to ambiguity disappeared, which may mean that all ambiguity of STAR result come from overlapping.
#####Another point is the number of unassigned reads with no features, which in both graph are the same, maybe this property is due to other settings of featureCount but not a "variable" outcome by changing mode of featureCount from gene to exon. I guess commands like "--minOverlap" may influence this part.